lib/logstash/inputs/tcp.rb

```diff
@@ -222,9 +180,72 @@ def read(socket)
     socket.sysread(16384)
   end
 
+  def ssl_context
+    return @ssl_context if @ssl_context
+
+    begin
+      @ssl_context = OpenSSL::SSL::SSLContext.new
+      @ssl_context.cert = OpenSSL::X509::Certificate.new(File.read(@ssl_cert))
+      @ssl_context.key = OpenSSL::PKey::RSA.new(File.read(@ssl_key),@ssl_key_passphrase)
+      if @ssl_verify
+        @cert_store = OpenSSL::X509::Store.new
+        # Load the system default certificate path to the store
+        @cert_store.set_default_paths
+        if File.directory?(@ssl_cacert)
+          @cert_store.add_path(@ssl_cacert)
+        else
+          @cert_store.add_file(@ssl_cacert)
+        end
+        @ssl_context.cert_store = @cert_store
+        @ssl_context.verify_mode = OpenSSL::SSL::VERIFY_PEER|OpenSSL::SSL::VERIFY_FAIL_IF_NO_PEER_CERT
+      end
+    rescue => e
+      @logger.error("Could not inititalize SSL context", :exception => e, :backtrace => e.backtrace)
+      raise e
+    end
+
+    @ssl_context
+  end
+
+  def new_server_socket
+    @logger.info("Starting tcp input listener", :address => "#{@host}:#{@port}")
+
+    begin
+      socket = TCPServer.new(@host, @port)
+    rescue Errno::EADDRINUSE
+      @logger.error("Could not start TCP server: Address in use", :host => @host, :port => @port)
+      raise
+    end
+
+    @ssl_enable ? OpenSSL::SSL::SSLServer.new(socket, ssl_context) : socket
+  end
+
+  def new_client_socket
+    socket = TCPSocket.new(@host, @port)
+
+    if @ssl_enable
+      socket = OpenSSL::SSL::SSLSocket.new(socket, ssl_context)
+      socket.connect
+    end
+
+    @logger.debug? && @logger.debug("Opened connection", :client => "#{socket.peer}")
+
+    socket
+  rescue OpenSSL::SSL::SSLError => e
+    @logger.error("SSL Error", :exception => e, :backtrace => e.backtrace)
+    # catch all rescue nil on close to discard any close errors or invalid socket
+    socket.close rescue nil
+    sleep(1) # prevent hammering peer
+    retry
```

Comment:
hmm, good question, this was the previous logic I reproduced here. I believe the original intent was  to try and connect ad-nauseam hoping the other side would eventually get fixed. 

Also, if the plugin terminates, the pipeline will restart it so in the end it will end up being the same. Any particular reason exiting to the pipeline and restarting would be preferable?